diffcp/cone_program.py

```diff
@@ -163,44 +181,90 @@ def derivative(dA, db, dc, **kwargs):
+            ])
+            solve = splinalg.factorized(M_iref.tocsc())
+            rhs = MT @ rhs
+            dz = np.zeros(N)
+            # iterative refinement
+            for _ in range(it_ref_iters):
+                residual = rhs - MT @ (M @ dz)
+                if np.linalg.norm(residual) <= 1e-10:
+                    break
+                dz = dz + solve(np.append(np.zeros(N), residual))[N:]
+        elif mode == "lsqr":
             dz = _solve_derivative(Q, cones_parsed, u, v, w, rhs)
+        else:
+            raise NotImplementedError(f'Unrecognized mode: {mode}')
+
         du, dv, dw = np.split(dz, [n, n + m])
         dx = du - x * dw
-        dy = D_proj_dual_cone.matvec(dv) - y * dw
-        ds = D_proj_dual_cone.matvec(dv) - dv - s * dw
+        if mode in ['dense', 'sparse']:
+            dy = D_proj_dual_cone @ dv - y * dw
+            ds = D_proj_dual_cone @ dv - dv - s * dw
+        else:
+            dy = D_proj_dual_cone.matvec(dv) - y * dw
+            ds = D_proj_dual_cone.matvec(dv) - dv - s * dw
         return -dx, -dy, -ds
 
     def adjoint_derivative(dx, dy, ds, **kwargs):
         """Applies adjoint of derivative at (A, b, c) to perturbations dx, dy, ds
-
         Args:
             dx: NumPy array representing perturbation in `x`
             dy: NumPy array representing perturbation in `y`
             ds: NumPy array representing perturbation in `s`
-
         Returns:
             (`dA`, `db`, `dc`), the result of applying the adjoint to the
             perturbations; the sparsity pattern of `dA` matches that of `A`.
         """
         dw = -(x @ dx + y @ dy + s @ ds)
-        dz = np.concatenate(
-            [dx, D_proj_dual_cone.rmatvec(dy + ds) - ds, np.array([dw])])
+        if mode == "dense" or mode == "sparse":
+            dz = np.concatenate(
+                [dx, D_proj_dual_cone.T @ (dy + ds) - ds, np.array([dw])])
+        else:
+            dz = np.concatenate(
+                [dx, D_proj_dual_cone.rmatvec(dy + ds) - ds, np.array([dw])])
+
         if np.allclose(dz, 0):
             r = np.zeros(dz.shape)
-        else:
+        elif mode == "dense":
+            r = np.linalg.lstsq(MT.todense(), dz, rcond=None)[0]
+        elif mode == "sparse":
+            rho = kwargs.get("rho", 1e-6)
+            it_ref_iters = kwargs.get("it_ref_iters", 5)
+            MT_iref = sparse.bmat([
+                [-sparse.eye(N), MT],
+                [M, rho * sparse.eye(N)]
+            ])
+            solve = splinalg.factorized(MT_iref.tocsc())
+            rhs = M @ dz
+            r = np.zeros(N)
+            # iterative refinement
+            for k in range(it_ref_iters):
+                residual = rhs - M @ (MT @ r)
+                if np.linalg.norm(residual) <= 1e-10:
+                    break
+                r = r + solve(np.append(np.zeros(N), residual))[N:]
+        elif mode == "lsqr":
             r = _solve_adjoint_derivative(Q, cones_parsed, u, v, w, dz)
```

Comment:
What do you think about renaming  `_solve{,_adjoint}_derivative` to something like `_solve{,_adjoint}_derivative_lsqr`?

Reply:
(Ah, just noticed you changed this in #11)